internal/internal_task_handlers.go

```diff
@@ -841,7 +828,7 @@ func (w *workflowExecutionContextImpl) ProcessWorkflowTask(workflowTask *workflo
 	var replayCommands []*commandpb.Command
 	var respondEvents []*historypb.HistoryEvent
 
-	msgs := indexMessages(workflowTask)
+	msgs := indexMessagesByEventID(workflowTask.task.GetMessages())
```

Comment:
Would it have value to ask the server to always deliver these in event ID order?

Also, is the number of these so great that you really need to sort because of fear of the cost of `takeLTE`? I think you can just make `takeLTE` iterate over all messages every time and https://github.com/golang/go/wiki/SliceTricks#filter-in-place to keep the ones greater than while extracting the ones less than or equal.

So something like this (untested, just typed here in comments):

```go
func (emi *eventMsgIndex) takeLTE(eventID int64) (out []*protocolpb.Message) {
	n := 0
	for _, msg := range *emi {
		if msg.GetEventID() > eventID {
			(*emi)[n] = msg
			n++
		} else {
			out = append(out, msg)
		}
	}
	*emi = *emi[:n]
	return;
}
```

Reply:
I'm fine with doing a little more work in takeLTE. I would expect the number of messages per WFT to be low (which to be fair, also makes sorting cheap) but there may be a substantial number when we rebundle into a bigger WFT for replay.